### ISSUE TITLE
Serde serialization vs consensus serialization. 

### ISSUE THREAD

**Issue by sanket1729** [rust-bitcoin/rust-bitcoin#765]
I think @thomaseizinger makes a fair point in  https://github.com/rust-bitcoin/rust-bitcoin/pull/756#discussion_r780013937. Basically, should we try to make sure for all data structures that serde serialization is the same as network serialization? It seems good to have this, but we lose the convenience of auto derives, have more code bloat that is annoying to write and review. 

I think we should just bite the bullet and document the serde feature saying that please don't confuse this with network serialization.

**Comment by Kixunil** [comment]
We could also write macro(s) implementing serde with consensus encode/decode.

**Comment by Kixunil** [comment]
My own opinion is that non-human-friendly formats should use binary consensus encoding while human-friendly should use common representations (hex for Txids and such, base56/bech32[m] for addresses, base64 for PSBTs...) and even structure (e.g. `Transaction` as object)

However if we change anything it'll be probably painful to upgrade for people already using serde. I think that at leas we should figure out some upgrade path. (E.g. release at least one version with both old and new style.)

**Comment by thomaseizinger** [comment]
Reposting my comment from [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/756#discussion_r780860596):

If all the fields are public, what is the value of providing these derives? If I need a data structure in my program that represents `EcdsaSig`, I can just add one myself and automatically get the freedom of choosing the field names I want etc.

I think we should seriously consider not providing serde implementations for these kind of container structs where there isn't a defined serialization format.

It triggers too many questions like "Is the defined serde format part of the public API?" Fixing a compile error because a field got renamed is one thing. Hitting a runtime deserialization error because a previously serialized struct changed a field name is much more annoying and invisible unless the end user adds explicit tests for it.

Additionally, for things like `Transaction`, I'd actually appreciate if we could expose a serde module that serializes to the consensus encoding: https://github.com/itchysats/itchysats/blob/58434af91d9692439a42f53f8b92a9d8b9193d72/daemon/src/model/cfd.rs#L1628-L1649

IMO, the need for such a module highlights that while choosing to derive `Serialize` might be convenient, it is also very opinionated. I think users would be better served by not implementing `Serialize` by default and instead provide multiple modules that users can use via `#[serde(with)]`:

- A module that serializes to consensus encoding like the one linked above
- A module that serializes to the internal structure, which can be created using serde's `remote` feature: https://serde.rs/remote-derive.html

> However if we change anything it'll be probably painful to upgrade for people already using serde.

The upgrade would be smooth if we provide a module that represents the current serialization.

> but we lose the convenience of auto derives, have more code bloat that is annoying to write and review.

I think this is also solved by serde's `remote` derive feature.

**Comment by apoelstra** [comment]
I'm a bit confused what "serialize to the consensus encoding" should mean for serde derives. The serializer defines the encoding, all we can do is tell the serializer what types certain objects present as.

I tesd to agree with @thomaseizinger about serde serializations tying our hands regarding field renamings and reorderings, in a subtle/scary way. We can add serde regression tests for everything (see #724) but that's a fairly big project that can't be guaranteed to catch everything. Whether we should conclude "no serde derives for convenience-container types"...that's not an easy question.

We could have a rule that we implement serde for container types, but they always need to de/serialize as tuples, and in some fixed/canonical order? I think this wouldn't tie our hands too much, nor be too difficult to implement (worst case we could macro it up).

**Comment by Kixunil** [comment]
I mean we could always write `serialize_bytes(consensus_encoded_bytes)`/`visit_bytes()`. If the serializer decides to transform it in weird ways, it's not our problem.

I believe many types should serialize this way for **non-human-friendly** formats.

**Comment by apoelstra** [comment]
It would be quite invasive to do this without extraneous allocations, but ok, I agree in principle.

Though I don't see how a serializer can avoid "transforming it in weird ways". At the very least it would need to length-prefix the data because it doesn't know how to interpret self-describing Bitcoin datatypes.

**Comment by Kixunil** [comment]
Yes and it's fine since it's expected. People don't complain about e.g. Json adding brackets around arrays.

Allocations are unfortunate, not sure if there's a sensible way around them.

**Comment by apoelstra** [comment]
It's expected, but it's still not consensus encoding, and it woud be less efficient than either consensus encoding or the current serde encoding, so I'm not sure what the point would be.

**Comment by apoelstra** [comment]
In cases where the `derive`s are non-compact then yes, we probably should replace them. But directly using consensus encoding would not be optimal (because of length prefixes etc). I don't think there are many cases of this.

Regarding stability, we should have regression tests and avoid this sort of breakage, yes, which may mean switching off of derives if we need to rename/reorder struct fields. I could also see an argument for avoiding derives entirely, if we think we could more naturally encoding things without field names, or with different field names/orderings.

But my point is that any talk about consensus encoding is a red herring, at best. We cannot get consensus encoding out of serde, and if we use our consensus encoding logic with serde the result will be a reduction in space efficiency (and a reduction in performance without extensive changes to our consensus encoding logic).

**Comment by Kixunil** [comment]
I'm not claiming consensus encoding is optimal. It's just much, much more efficient than e.g. stringly Json. I think that using it is good trade off. A nice thing about it is you could have e.g. two applications that exchange structured messages. Sender encodes a transaction and some metadata and the receiver doesn't decode the transaction, just processes the metadata and sends the already-encoded transaction straight to the network with minimal "parsing" (just compute offsets/sizes and take the subslice).

My idea is that consensus encoding is the most widely used somewhat efficient encoding.

**Comment by apoelstra** [comment]
I don't understand where we are talking past each other :(. serde does not give us any way to detect if the user is using json and to prevent them from doing so. Nor does it give us any way to use consensus encoding, no matter what encoder the user has chosen.

These limitations of serde are exactly why we have the consensus encoding/decoding traits to begin with.

**Comment by Kixunil** [comment]
Oh, maybe you're unaware of this function? https://docs.rs/serde/latest/serde/trait.Serializer.html#method.is_human_readable

**Comment by apoelstra** [comment]
I am aware of this function, and have used it in pretty-much every manual serde implementation I've written. I continue to not understand what is wrong with the existing serde situation, what is meant by moving it "closer to consensus encoding", or what this would improve.

**Comment by thomaseizinger** [comment]
Perhaps I should have been more clear in my comment above.

I never intended to suggest that we _replace_ the consensus serialization code with serde. That is hard/impossible as @apoelstra highlighted because serde has its own data format and decouples from the actual encoding which is unnecessary for Bitcoin's format.

I think the way we should approach using serde for types that have a consensus encoding is that we should delegate to it and emit a single, opaque value to serde:

- a hex-string if it is human readable
- a byte array if it is not

If that being the default is considered too opinionated, then I'd appreciate a module for it. (But we appear to already have consensus - _pun intended_ - on that.)

**Comment by apoelstra** [comment]
This sounds fine to me, though it would be nice if it were possible to do without allocating temporary buffers.

**Comment by thomaseizinger** [comment]
> This sounds fine to me, though it would be nice if it were possible to do without allocating temporary buffers.

We could encode into an array that lives on the stack. `&mut [u8]` implements `io::Write`. Still a buffer but at least not allocated on the heap?

For hex, it might be a bit more involved. We would probably need another byte array (twice the length), convert to hex into that and use https://doc.rust-lang.org/std/str/fn.from_utf8.html to get a `&str` which we can pass to serde.

So I think the answer is, yes we should be able to do it without allocating.

**Comment by Kixunil** [comment]
Yeah, I didn't mean to suggest replacing  consensus encode with serde either.

One issue with allocating on stack is how large we want to make the array. Nothing in `blockdata` should be larger than 4MB, but 4MB may still be too high on embedded platforms? I will ask about adding iterator support to serde.

**Comment by apoelstra** [comment]
I think `serialize_seq` will result in JSON serialization of arrays (and possibly analogous things in binary encodings) rather than as single blocks of bytes/hex chars.

**Comment by Kixunil** [comment]
Json will be hex where we can use `collect_str`

**Comment by apoelstra** [comment]
We are betting then that only human-readable encoders will do crazy inefficient things in response to `serialize_seq`, and that non-human-readable encoders will do efficient things. I'm really hesitant to make that bet.

**Comment by Kixunil** [comment]
OK, will ask at `serde`

**Comment by Kixunil** [comment]
https://github.com/serde-rs/serde/issues/2155

**Comment by Kixunil** [comment]
Since that one was closed as wontfix, I suggest to check what popular non-human-readable encoders (bincode, what else?) do and rely on `serialize_seq` if it's beneficial to them.

**Comment by tcharding** [comment]
Woah, `consensus/serde.rs` is a hell of a file. I'm not sure I can add any meaningful review except to ask questions :) I did have a good play with it, perhaps we can have this unit test (I added a `pub use hex::{Lower, Upper}` also)
```rust

#[cfg(test)]
mod tests {
    use crate::Transaction;
    use crate::consensus::encode::deserialize;
    use crate::consensus::serde::{Hex, Upper};
    use crate::hashes::hex::FromHex;

    #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
    #[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
    pub struct MyStruct {
        #[serde(with = "consenus::serde::With::<Hex<Upper>>")]
        tx: Transaction,
    }

    #[test]
    fn serialize_my_struct_using_consensus_uppercase_hex() {
        let tx_hex = "010000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff3603da1b0e00045503bd5704c7dd8a0d0ced13bb5785010800000000000a636b706f6f6c122f4e696e6a61506f6f6c2f5345475749542fffffffff02b4e5a212000000001976a914876fbb82ec05caa6af7a3b5e5a983aae6c6cc6d688ac0000000000000000266a24aa21a9edf91c46b49eb8a29089980f02ee6b57e7d63d33b18b4fddac2bcd7db2a39837040120000000000000000000000000000000000000000000000000000000000000000000000000";
        let tx_bytes = Vec::from_hex(tx_hex).unwrap();
        let tx: Transaction = deserialize(&tx_bytes).expect("failed to deserialize tx bytes");

        let orig = MyStruct { tx };

        let want = format!("{{\"tx\":\"{}\"}}", tx_hex.to_uppercase());
        let got = serde_json::to_string(&orig).expect("failed to serialize json");

        assert_eq!(got, want)
    }
}
```

**Comment by apoelstra** [comment]
@tcharding I think you meant to post this on #1254

**Comment by tcharding** [comment]
Ha, you were fast - I deleted it about 15 seconds after I posted it :)
